round-based/src/state_machine/delivery.rs

```diff
@@ -0,0 +1,88 @@
+use core::task::{ready, Poll};
+
+/// Stream of incoming messages
+pub struct Incomings<M> {
+    shared_state: super::shared_state::SharedStateRef<M>,
+}
+
+impl<M> Incomings<M> {
+    pub(super) fn new(shared_state: super::shared_state::SharedStateRef<M>) -> Self {
+        Self { shared_state }
+    }
+}
+
+impl<M> crate::Stream for Incomings<M> {
+    type Item = Result<crate::Incoming<M>, core::convert::Infallible>;
+
+    fn poll_next(
+        self: core::pin::Pin<&mut Self>,
+        _cx: &mut core::task::Context<'_>,
+    ) -> Poll<Option<Self::Item>> {
+        let scheduler = ready!(self.shared_state.can_schedule());
+
+        scheduler
+            .protocol_needs_more_msgs()
+            .map(|msg| Some(Ok(msg)))
+    }
+}
+
+/// Sink for outgoing messages
+pub struct Outgoings<M> {
+    shared_state: super::shared_state::SharedStateRef<M>,
+}
+
+impl<M> Outgoings<M> {
+    pub(super) fn new(shared_state: super::shared_state::SharedStateRef<M>) -> Self {
+        Self { shared_state }
+    }
+}
+
+impl<M> crate::Sink<crate::Outgoing<M>> for Outgoings<M> {
+    type Error = SendErr;
+
+    fn poll_ready(
+        self: core::pin::Pin<&mut Self>,
+        _cx: &mut core::task::Context<'_>,
+    ) -> Poll<Result<(), Self::Error>> {
+        let scheduler = ready!(self.shared_state.can_schedule());
+        scheduler.protocol_flushes_outgoing_msg().map(Ok)
+    }
+
+    fn start_send(
+        self: core::pin::Pin<&mut Self>,
+        msg: crate::Outgoing<M>,
+    ) -> Result<(), Self::Error> {
+        self.shared_state
+            .protocol_saves_msg_to_be_sent(msg)
+            .map_err(|_| SendErr(SendErrReason::NotReady))
+    }
```

Comment:
It wasn't obvious to me that this would work with `SinkExt::send_all`, but it does because of the flushing in `poll_ready`. Would be good to add comments and maybe tests. I can send you a diff with the test I wrote.

Reply:
BTW we should use send_all in our protocols

Reply:
> BTW we should use send_all in our protocols

Why should we ever use it? We never have a stream of messages to send

Reply:
> We never have a stream of messages to send

We do all the time when doing p2p messages. It's not a stream but rather an iterator, but this library has no method specifically for iterators, but it does have `futures::stream::iter`.

> if it’s any better

In case a sink supports it, it allows us to buffer and send them together instead of doing `n` network transfers

Reply:
> It's not a stream but rather an iterator, but this library has no method specifically for iterators, but it does have futures::stream::iter.

feed/flush should work perfectly fine

Reply:
> In case a sink supports it, it allows us to buffer and send them together instead of doing n network transfers

Yeah but the overall performance might be better if each party does a network transfer right away when it has a message. When parties buffer a message instead of sending, that means that other party will receive the buffered message later than it could, intuitively it will cause the overall performance decrease

Reply:
> buffer a message instead of sending

It seems we have different situations in mind. In ku23, the messages are already computed before sending, so there is no noticable wait between adding them to the queue, we lose more time waiting for delivery to ack our message. So buffering would help here, by reducing the amount of roundtrips between the party and the delivery. But I agree, in other protocols where we spend tens of milliseconds to create a message, it's better to send it immediately

Reply:
Yeah, in KU23 it might be better to do that indeed. I was thinking about cggmp21 where there's a noticeable delay between sending messages

Reply:
It might be more ergonomic to use `send_all` in ku23 as well...